src/blockchain/node/filteroracle/BlockIndexer.hpp

```diff
@@ -50,141 +37,95 @@ class Session;
 
 namespace blockchain
 {
-namespace bitcoin
-{
-namespace block
-{
-class Block;
-}  // namespace block
-}  // namespace bitcoin
-
 namespace database
 {
 class Cfilter;
 }  // namespace database
 
 namespace node
 {
-namespace implementation
-{
 class FilterOracle;
-}  // namespace implementation
-
-namespace internal
-{
-class BlockOracle;
 class Manager;
-}  // namespace internal
-
-class HeaderOracle;
 }  // namespace node
 }  // namespace blockchain
 
 namespace network
 {
 namespace zeromq
 {
-namespace socket
-{
-class Push;
-}  // namespace socket
-
 class Message;
 }  // namespace zeromq
 }  // namespace network
 // }  // namespace v1
 }  // namespace opentxs
 // NOLINTEND(modernize-concat-nested-namespaces)
 
-namespace opentxs::blockchain::node::implementation
+namespace opentxs::blockchain::node::filteroracle
 {
-using BlockDMFilter = download::Manager<
-    FilterOracle::BlockIndexer,
-    std::shared_ptr<const bitcoin::block::Block>,
-    cfilter::Header,
-    cfilter::Type>;
-using BlockWorkerFilter = Worker<api::Session>;
-
-class FilterOracle::BlockIndexer final : public BlockDMFilter,
-                                         public BlockWorkerFilter
+class BlockIndexer::Imp final : public Actor<BlockIndexerJob>
 {
 public:
-    auto NextBatch() noexcept { return allocate_batch(type_); }
+    auto Init(boost::shared_ptr<Imp> me) noexcept -> void;
+    auto Reindex() noexcept -> void;
+    auto Shutdown() noexcept -> void;
 
-    BlockIndexer(
-        const api::Session& api,
+    Imp(const api::Session& api,
+        const node::Manager& node,
+        const node::FilterOracle& parent,
         database::Cfilter& db,
-        const HeaderOracle& header,
-        const internal::BlockOracle& block,
-        const internal::Manager& node,
-        FilterOracle& parent,
-        const blockchain::Type chain,
-        const cfilter::Type type,
-        const UnallocatedCString& shutdown,
-        const NotifyCallback& notify) noexcept;
-
-    ~BlockIndexer() final;
-
-protected:
-    auto pipeline(zmq::Message&& in) -> void final;
-    auto state_machine() noexcept -> bool final;
+        NotifyCallback&& notify,
+        blockchain::Type chain,
+        cfilter::Type type,
+        std::string_view parentEndpoint,
+        const network::zeromq::BatchID batch,
+        allocator_type alloc) noexcept;
+    Imp() = delete;
+    Imp(const Imp&) = delete;
+    Imp(Imp&&) = delete;
+    auto operator=(const Imp&) -> Imp& = delete;
+    auto operator=(Imp&&) -> Imp& = delete;
+
+    ~Imp() final;
 
 private:
-    auto shut_down() noexcept -> void;
+    friend Actor<BlockIndexerJob>;
```

Comment:
W podklasach przeniesionych z Worker do Actor nie ma dobrego powodu używać friend Actor<X>. Jeśli bez tego friend relationship się nie kompiluje, to coś wymaga dalszej poprawki, prawdopodobnie wprowadzenia override albo final dla jednej z funkcji czysto wirtualnych Actor.

Reply:
zrobione